provider/servicefabric/servicefabric.go

```diff
@@ -0,0 +1,418 @@
+	return nil
+}
+
+func (p *Provider) updateConfig(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, sfClient sfsdk.Client, pollInterval time.Duration) error {
+	pool.Go(func(stop chan bool) {
+		operation := func() error {
+			ticker := time.NewTicker(pollInterval)
+			for range ticker.C {
+				select {
+				case shouldStop := <-stop:
+					if shouldStop {
+						ticker.Stop()
+						return nil
+					}
+				default:
+					log.Info("Checking service fabric config")
+				}
+
+				services, err := p.getClusterServices(sfClient)
+				if err != nil {
+					log.Error(err)
+					return err
+				}
+
+				templateObjects := struct {
+					Services []ServiceItemExtended
+				}{
+					services,
+				}
+
+				var sfFuncMap = template.FuncMap{
+					"isPrimary":                       p.isPrimary,
+					"isHealthy":                       p.isHealthy,
+					"hasHTTPEndpoint":                 p.hasHTTPEndpoint,
+					"getDefaultEndpoint":              p.getDefaultEndpoint,
+					"getNamedEndpoint":                p.getNamedEndpoint,
+					"getApplicationParameter":         p.getApplicationParameter,
+					"doesAppParamContain":             p.doesAppParamContain,
+					"hasServiceLabel":                 p.hasServiceLabel,
+					"getServiceLabelValue":            p.getServiceLabelValue,
+					"getServiceLabelValueWithDefault": p.getServiceLabelValueWithDefault,
+					"getServiceLabelsWithPrefix":      p.getServiceLabelsWithPrefix,
+					"getServicesWithLabelValueMap":    p.getServicesWithLabelValueMap,
+					"getServicesWithLabelValue":       p.getServicesWithLabelValue,
+				}
+
+				configuration, err := p.GetConfiguration("templates/servicefabric.tmpl", sfFuncMap, templateObjects)
+
+				if err != nil {
+					log.Error(err)
+				}
+
+				configurationChan <- types.ConfigMessage{
+					ProviderName:  "servicefabric",
+					Configuration: configuration,
+				}
+			}
+			return nil
+		}
+		notify := func(err error, time time.Duration) {
+			log.Errorf("Provider connection error: %s; retrying in %s", err, time)
+		}
+		err := backoff.RetryNotify(safe.OperationWithRecover(operation), job.NewBackOff(backoff.NewExponentialBackOff()), notify)
+		if err != nil {
+			log.Errorf("Cannot connect to Provider: %s", err)
+		}
+	})
+	return nil
+}
+
+func (p *Provider) getClusterServices(sfClient sfsdk.Client) ([]ServiceItemExtended, error) {
+	results := []ServiceItemExtended{}
+	apps, err := sfClient.GetApplications()
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+	for _, app := range apps.Items {
+		services, err := sfClient.GetServices(app.ID)
+		if err != nil {
+			log.Error(err)
+			return nil, err
+		}
+
+		for _, service := range services.Items {
+			item := ServiceItemExtended{
+				ServiceItem: service,
+				Application: app,
+				Labels:      make(map[string]string),
+			}
+
+			addLabelsFromServiceExtension(sfClient, service.TypeName, &app, &item)
+			addLabelsFromPropertyManager(sfClient, &item)
```

Comment:
Could be directly handled in sfClient ?

Reply:
These calls wrap the underlying sfsdk calls GetServiceExtension and GetProperties we've aimed to pull as much out as we can while keeping the sdk generalized not traefik specific. 

`addLabelsFromPropertyManager` deserializes into an XML type, we've moved as much of the code as we can into the sdk but the type is specific to our usage in traefik. 

`addLabelsFromPropertyManager` is a very thin wrapper which filters the properties to only those with a `traefik.` prefix. 

Happy to take advice on different ways of handling this, hopefully we've ended up at a reasonable trade off